zio/csvio/reader.go

```diff
@@ -0,0 +1,109 @@
+package csvio
+
+import (
+	"encoding/csv"
+	"errors"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zson"
+)
+
+type Reader struct {
+	zctx      *zson.Context
+	reader    *csv.Reader
+	marshaler *zson.MarshalZNGContext
+	strings   bool
+	valid     bool
+	hdr       []string
+	vals      []interface{}
+}
+
+// XXX This is a placeholder for an option that will allow one to convert
+// all csv fields to strings and defer any type coercion presumably to
+// Z shapers.  Currently, this causes an import cycle because the csvio
+// Writer depends on fuse.  We should refactor this so whatever logic wants
+// to tack on a fuse operator happens outside of zio.  See issue #2315
+//type ReaderOpts struct {
+//	StringsOnly bool
+//}
+
+func NewReader(reader io.Reader, zctx *zson.Context) *Reader {
+	return &Reader{
+		zctx:      zctx,
+		reader:    csv.NewReader(reader),
+		marshaler: zson.NewZNGMarshaler(),
+		//strings:   opts.StringsOnly,
+	}
+}
+
+func (r *Reader) Read() (*zng.Record, error) {
+	for {
+		csvRec, err := r.reader.Read()
+		if err != nil {
+			if err == io.EOF {
+				if !r.valid {
+					err = errors.New("empty csv file")
```

Comment:
I assume this here to get auto-detect working? I would assume that expected behavior for a Reader on an empty file would be a `nil, nil` return on the first read...

Reply:
Yes, that was my first crack at getting the auto-detector to work then I realized it was more involved because CSV is so forgiving to short, random inputs.  I could take this out and it could be part of the "strict mode" that is contemplated in #2319 2316.

Reply:
I'm inclined to leave it for now and address this in #2316